crates/extensions/c8y_firmware_manager/src/actor.rs

```diff
@@ -0,0 +1,723 @@
+
+        match FirmwareOperationResponse::try_from(&message) {
+            Ok(response) => {
+                if let Err(err) = self
+                    .handle_child_device_firmware_update_response(&response, message_box)
+                    .await
+                {
+                    self.fail_operation_in_cloud(
+                        &child_id,
+                        Some(response.get_payload().operation_id.as_str()),
+                        &err.to_string(),
+                        message_box,
+                    )
+                    .await?;
+                }
+            }
+            Err(err) => {
+                // Ignore bad responses. Eventually, timeout will fail an operation.
+                error!("Received a firmware update response with invalid payload for child {child_id}. Error: {err}");
+            }
+        }
+        Ok(())
+    }
+
+    async fn handle_child_device_firmware_update_response(
+        &mut self,
+        response: &FirmwareOperationResponse,
+        message_box: &mut FirmwareManagerMessageBox,
+    ) -> Result<(), FirmwareManagementError> {
+        let child_device_payload = response.get_payload();
+        let child_id = response.get_child_id();
+        let operation_id = child_device_payload.operation_id.as_str();
+        let received_status = child_device_payload.status;
+        info!("Firmware update response received. Details: id={operation_id}, child={child_id}, status={received_status:?}");
+
+        let operation_key = OperationKey::new(&child_id, &operation_id);
+        let current_operation_state = self.active_child_ops.get(&operation_key);
+
+        match current_operation_state {
+            Some(&ActiveOperationState::Executing) => {}
+            Some(&ActiveOperationState::Pending) => {
+                self.publish_c8y_executing_message(&child_id, message_box)
+                    .await?;
+                // TODO: Maybe cutting off as a function.
+                self.active_child_ops
+                    .insert(operation_key.clone(), ActiveOperationState::Executing);
+            }
+            None => {
+                info!("Received a response from {child_id} for unknown request {operation_id}.");
+                return Ok(());
+            }
+        }
+
+        match received_status {
+            OperationStatus::Successful => {
+                let status_file_path = self.config.firmware_dir.join(operation_id);
+                let operation_entry =
+                    FirmwareOperationEntry::read_from_file(status_file_path.as_path())?;
+
+                self.publish_c8y_installed_firmware_message(&operation_entry, message_box)
+                    .await?;
+                self.publish_c8y_successful_message(&child_id, message_box)
+                    .await?;
+
+                self.remove_status_file(operation_id)?;
+                self.remove_entry_from_active_operations(&operation_key);
```

Comment:
This could be refactored into a single cleanup function which is used here as well as when the operation is failed.